#*Case Study Data Reqruitments*

# A.Fitur Extract

In [242]:
import pandas as pd
import pytz
from datetime import datetime
import re #untuk regex findall

In [243]:
from google.colab import drive
drive.mount('/content/drive')

Drive already mounted at /content/drive; to attempt to forcibly remount, call drive.mount("/content/drive", force_remount=True).


Fitur extract

In [244]:
def exctractDataCsv(path):
    data1 = pd.read_csv(path)

    #indek dihapus dari tampilan
    blanked_index=[''] * len(data1)
    data1.index=blanked_index

    return data1

data1 = exctractDataCsv('/content/drive/My Drive/purwadhika_files/data_sources/data_reqruitment/data_requirements.csv')
data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1555 entries,  to 
Data columns (total 14 columns):
 #   Column            Non-Null Count  Dtype  
---  ------            --------------  -----  
 0   Unnamed: 0        1555 non-null   int64  
 1   company           1551 non-null   object 
 2   company_rating    1358 non-null   float64
 3   location          1554 non-null   object 
 4   job_title         1554 non-null   object 
 5   job_description   1554 non-null   object 
 6   salary_estimate   1277 non-null   object 
 7   company_size      1442 non-null   object 
 8   company_type      1442 non-null   object 
 9   company_sector    1260 non-null   object 
 10  company_industry  1260 non-null   object 
 11  company_founded   1131 non-null   float64
 12  company_revenue   1442 non-null   object 
 13  dates             1555 non-null   object 
dtypes: float64(2), int64(1), object(11)
memory usage: 182.2+ KB


# B.Fitur Transformation

## 1.Drop Kolom Unnamed: 0

In [245]:
data1.drop('Unnamed: 0',axis=1,inplace=True) #menghapus kolom unnamed: 0 karena tdk diperlukan
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description     salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70,000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...   $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...   $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                 NaN   
  Description of Work:\nTechnical experience in ...   $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                      dates  
  Unknown / Non-Applicable  2024-06-12 00:00:00-10:00  
                       NaN  2024-06-12 00:00:00+07:00  
  Unknown / Non-Applicable  2024-06-12 00:00:00-10:00  
        $10+ billion (USD)  2024-06-12 00:00:00-05:00  
  Unknown / Non-Applicable  2024-06-12 00:00:00-04:00

## 2.Handle Null Values

In [246]:
data1.isna().sum() #cek apakah ada nilai null.

company               4
company_rating      197
location              1
job_title             1
job_description       1
salary_estimate     278
company_size        113
company_type        113
company_sector      295
company_industry    295
company_founded     424
company_revenue     113
dates                 0
dtype: int64

In [247]:
#hampir semua kolom memiliki null value, yang akan saya hapus hanya row yang company nya null, karena tidak valid, untuk kolom lainnya masih bisa diambil kesimpulan jika nilainya null.
data1.dropna(subset=['company'], inplace=True)
data1.isna().sum()

company               0
company_rating      193
location              0
job_title             0
job_description       0
salary_estimate     277
company_size        112
company_type        112
company_sector      294
company_industry    294
company_founded     423
company_revenue     112
dates                 0
dtype: int64

## 3.Ubah data type Dates menjadi datetime type dan seragamkan timezone

cek data type terlebih dahulu

In [248]:
data1.dtypes

company              object
company_rating      float64
location             object
job_title            object
job_description      object
salary_estimate      object
company_size         object
company_type         object
company_sector       object
company_industry     object
company_founded     float64
company_revenue      object
dates                object
dtype: object

mengubah kolom dates dari semula dtypes object menjadi dtypes Datetime, dan menyeragamkan timezone menjadi UTC+0

In [249]:
data1['dates'] = pd.to_datetime(data1['dates'], format='%Y-%m-%d %H:%M:%S%z', utc=True)
data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1551 entries,  to 
Data columns (total 13 columns):
 #   Column            Non-Null Count  Dtype              
---  ------            --------------  -----              
 0   company           1551 non-null   object             
 1   company_rating    1358 non-null   float64            
 2   location          1551 non-null   object             
 3   job_title         1551 non-null   object             
 4   job_description   1551 non-null   object             
 5   salary_estimate   1274 non-null   object             
 6   company_size      1439 non-null   object             
 7   company_type      1439 non-null   object             
 8   company_sector    1257 non-null   object             
 9   company_industry  1257 non-null   object             
 10  company_founded   1128 non-null   float64            
 11  company_revenue   1439 non-null   object             
 12  dates             1551 non-null   datetime64[ns, UTC]
dtypes: datetime64

In [250]:
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description     salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70,000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...   $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...   $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                 NaN   
  Description of Work:\nTechnical experience in ...   $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
                       NaN 2024-06-11 17:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
        $10+ billion (USD) 2024-06-12 05:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00

## 4.Ubah salary estimate menjadi float dan pastikan nilainya seragam (currency dan satuannya)

In [251]:
#cek apakah semua salary estimate dalam $
data1['salary_estimate'].str[0].unique()

array(['$', nan], dtype=object)

berarti hasil di atas menunjukkn semua
salary estimate dalam $

In [252]:
#cek salary estimate /hr dan /yr saja atau tidak
data1['salary_estimate'].str.extract(r' /(\w+) ')[0].unique()

array(['yr', 'hr', nan, 'mo'], dtype=object)

hasil di atas menunjukkan salary estimate satuannya hr(hour), yr(year), dan mo (month)

In [253]:
#replace comma menjadi ''
data1['salary_estimate'] = data1['salary_estimate'].str.replace(',','')
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
                       NaN 2024-06-11 17:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
        $10+ billion (USD) 2024-06-12 05:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00

In [254]:
#extract nilai salary
data1['salary_estimate'].str.extract(r'(\d+.\d+)')[0]
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
                       NaN 2024-06-11 17:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00  
        $10+ billion (USD) 2024-06-12 05:00:00+00:00  
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00

In [255]:
#buat kolom faktor kali est untuk standarisasi nilai salary est
data1['faktorkali_salary_est'] = data1['salary_estimate'].str.extract(r' /(\w+) ')[0]
data1['faktorkali_salary_est'] = data1['faktorkali_salary_est'].str.replace('yr','1848') #asumsi dalam 1 tahun rata2 work hour di US adalah 2087 jam. sumbee : https://www.indeed.com/career-advice/career-development/how-many-work-hours-in-a-year
data1['faktorkali_salary_est'] = data1['faktorkali_salary_est'].str.replace('hr','1')
data1['faktorkali_salary_est'] = data1['faktorkali_salary_est'].str.replace('mo','154') #asumsi rata2 jam kerja per bulan berdasar sumber jam kerja tahunan di atas dibagi 12
data1['faktorkali_salary_est']



1848
       1
       1
     NaN
       1
    ... 
    1848
    1848
    1848
    1848
    1848
Name: faktorkali_salary_est, Length: 1551, dtype: object

In [256]:
#fillna faktor kali salary est dg nilai 0
data1['faktorkali_salary_est'] = data1['faktorkali_salary_est'].fillna(0)
data1['faktorkali_salary_est']

1848
       1
       1
       0
       1
    ... 
    1848
    1848
    1848
    1848
    1848
Name: faktorkali_salary_est, Length: 1551, dtype: object

In [257]:
#membuat kolom std_salary_est_perhour
data1['value_salary_est'] = data1['salary_estimate'].str.extract(r'(\d+.\d+)')[0] #mencari value float dari extract string.
data1['value_salary_est'] = data1['value_salary_est'].astype(float) #mengubah datatype menjadi float agar bisa dioperasikan dg math operation
data1['std_salary_est_perhour'] = data1['value_salary_est'] / data1['faktorkali_salary_est'].astype(int) #math operation untuk standardise salary per hour in USD $
data1

company  company_rating           location  \
                 PCS Global Tech\n4.7             4.7      Riverside, CA   
           Futuretech Consultants LLC             NaN         Newton, MS   
                     Clairvoyant\n4.4             4.4             Remote   
                           Apple\n4.2             4.2      Cupertino, CA   
    Skytech Consultancy Services\n5.0             5.0      Baltimore, MD   
..                                ...             ...                ...   
                          Paycom\n3.5             3.5  Oklahoma City, OK   
      Amazon Data Services, Inc.\n3.7             3.7      Hermiston, OR   
              Palo Alto Networks\n4.2             4.2    Santa Clara, CA   
                        Talent Groups             NaN             Remote   
       Gainwell Technologies LLC\n3.4             3.4      Roseville, CA   

                                            job_title  \
                        Data Engineer | PAID BOOTCAMP   
                              Snowflake Data Engineer   
                                  Data Engineer (MDM)   
                                        Data Engineer   
                                        Data Engineer   
..                                                ...   
                                    Sr. Data Engineer   
    Data Center Regional Electrical Engineer (Fiel...   
    Sr Principal Data Software Engineer (SASE, Cloud)   
    Fulltime // Data Engineer (H4 EADs, L2 EADs, G...   
         Cloud Data Engineer - Hybrid - Roseville, CA   

                                      job_description     salary_estimate  \
    Responsibilities\n· Analyze and organize raw d...   $70000 /yr (est.)   
    My name is Dileep and I am a recruiter at Futu...   $42.50 /hr (est.)   
    Required Skills:\nMust have 5-8+ Years of expe...   $67.50 /hr (est.)   
    Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                 NaN   
    Description of Work:\nTechnical experience in ...   $65.00 /hr (est.)   
..                                                ...                 ...   
    This position will be located within the Devel...   $90582 /yr (est.)   
    Bachelor’s Degree in Electrical Engineering or...  $105893 /yr (est.)   
    Company Description\n\nOur Mission\nAt Palo Al...  $188700 /yr (est.)   
    Position: Data engineer (Live Table Structured...  $130000 /yr (est.)   
    Be part of a team that unleashes the power of ...   $98850 /yr (est.)   

              company_size       company_type                  company_sector  \
     501 to 1000 Employees  Company - Private          Information Technology   
                       NaN                NaN                             NaN   
       51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
          10000+ Employees   Company - Public          Information Technology   
         1 to 50 Employees   Company - Public                             NaN   
..                     ...                ...                             ...   
    1001 to 5000 Employees   Company - Public          Information Technology   
          10000+ Employees   Company - Public          Information Technology   
          10000+ Employees   Company - Public          Information Technology   
                   Unknown  Company - Private      Human Resources & Staffing   
          10000+ Employees  Company - Private          Information Technology   

                           company_industry  company_founded  \
    Information Technology Support Services              NaN   
                                        NaN              NaN   
                  Biotech & Pharmaceuticals              NaN   
              Computer Hardware Development           1976.0   
                                        NaN              NaN   
..                                      ...              ...   
    Enterprise Software & Network Solutions           1998.0   
                    In

## 5.Melakukan labelling pada kolom company_revenue

In [258]:
#cek apakah company revenue dalam dollar USD smua
data1['company_revenue'].str.extract(r'\((.*?)\)')[0].unique()

array([nan, 'USD'], dtype=object)

In [259]:
data1['company_revenue'].unique()

array(['Unknown / Non-Applicable', nan, '$10+ billion (USD)',
       '$5 to $25 million (USD)', 'Less than $1 million (USD)',
       '$5 to $10 billion (USD)', '$1 to $5 billion (USD)',
       '$25 to $100 million (USD)', '$1 to $5 million (USD)',
       '$100 to $500 million (USD)', '$500 million to $1 billion (USD)'],
      dtype=object)

dari hasil di atas , semua company revanue dalam USD.

In [260]:
#labeling company_revenue menjadi grade yang dapat diurutkan atau dilakukan agregrate
data1.loc[data1['company_revenue'] == 'Less than $1 million (USD)', 'grade_company_revanue'] = 1
data1.loc[data1['company_revenue'] == '$1 to $5 million (USD)', 'grade_company_revanue'] = 2
data1.loc[data1['company_revenue'] == '$5 to $25 million (USD)', 'grade_company_revanue'] = 3
data1.loc[data1['company_revenue'] == '$25 to $100 million (USD)', 'grade_company_revanue'] = 4
data1.loc[data1['company_revenue'] == '$100 to $500 million (USD)', 'grade_company_revanue'] = 5
data1.loc[data1['company_revenue'] == '$500 million to $1 billion (USD)', 'grade_company_revanue'] = 6
data1.loc[data1['company_revenue'] == '$1 to $5 billion (USD)', 'grade_company_revanue'] = 7
data1.loc[data1['company_revenue'] == '$5 to $10 billion (USD)', 'grade_company_revanue'] = 8
data1.loc[data1['company_revenue'] == '$10+ billion (USD)', 'grade_company_revanue'] = 9
data1.loc[data1['company_revenue'] == 'Unknown / Non-Applicable', 'grade_company_revanue'] = 0
data1.loc[data1['company_revenue'].isnull(),  'grade_company_revanue'] = 0

data1.head()


company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates faktorkali_salary_est  \
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                  1848   
                       NaN 2024-06-11 17:00:00+00:00                     1   
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                     1   
        $10+ billion (USD) 2024-06-12 05:00:00+00:00                     0   
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00                     1   

  value_salary_est  std_salary_est_perhour  grade_company_revanue  
           70000.0               37.878788                    0.0  
              42.5               42.500000                    0.0  
              67.5               67.500000                    0.0  
               NaN                     NaN                    9.0  
              65.0               65.000000                    0.0

In [261]:
data1['grade_company_revanue'] = data1['grade_company_revanue'].astype('int32')
data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1551 entries,  to 
Data columns (total 17 columns):
 #   Column                  Non-Null Count  Dtype              
---  ------                  --------------  -----              
 0   company                 1551 non-null   object             
 1   company_rating          1358 non-null   float64            
 2   location                1551 non-null   object             
 3   job_title               1551 non-null   object             
 4   job_description         1551 non-null   object             
 5   salary_estimate         1274 non-null   object             
 6   company_size            1439 non-null   object             
 7   company_type            1439 non-null   object             
 8   company_sector          1257 non-null   object             
 9   company_industry        1257 non-null   object             
 10  company_founded         1128 non-null   float64            
 11  company_revenue         1439 non-null   object       

In [262]:
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates faktorkali_salary_est  \
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                  1848   
                       NaN 2024-06-11 17:00:00+00:00                     1   
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                     1   
        $10+ billion (USD) 2024-06-12 05:00:00+00:00                     0   
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00                     1   

  value_salary_est  std_salary_est_perhour  grade_company_revanue  
           70000.0               37.878788                      0  
              42.5               42.500000                      0  
              67.5               67.500000                      0  
               NaN                     NaN                      9  
              65.0               65.000000                      0

## 6.Memisahkan Location menjadi City dan Country

In [263]:
data1['location'].unique()

array(['Riverside, CA', 'Newton, MS', 'Remote', 'Cupertino, CA',
       'Baltimore, MD', 'New York, NY', 'Westwood, MA',
       'San Francisco, CA', 'Alpharetta, GA', 'Irving, TX', 'Boston, MA',
       'Salt Lake City, UT', 'Braham, MN', 'East Hartford, CT',
       'Dallas, TX', 'Arlington, VA', 'Framingham, MA', 'Syracuse, NY',
       'Princeton, NJ', 'Washington, DC', 'Menlo Park, CA',
       'Redlands, CA', 'Austin, TX', 'Richmond, VA', 'Atlanta, GA',
       'Berkeley Heights, NJ', 'South Milwaukee, WI', 'Chicago, IL',
       'Richville, MN', 'McLean, VA', 'Oakland, CA', 'Glen Allen, VA',
       'Newark, NJ', 'Dekalb, IL', 'Beaverton, OR', 'Plainsboro, NJ',
       'Los Angeles, CA', 'Pittsburgh, PA', 'Mahwah, NJ', 'Murray, UT',
       'Santa Clara, CA', 'Plano, TX', 'Madison, WI', 'Mount Laurel, NJ',
       'Cleveland, OH', 'Minneapolis, MN', 'Santa Ana, CA',
       'Rockville, MD', 'Dover, DE', 'Brentwood, TN', 'Edison, NJ',
       'Jersey City, NJ', 'Hanover, NJ', 'Imperial, CA', 

In [264]:
#memisahkan city dari location
data1['city'] = data1['location'].str.split(',').str[0]

#memisahkan country dari location
data1['country'] = data1['location'].str.split(',').str[1]
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates faktorkali_salary_est  \
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                  1848   
                       NaN 2024-06-11 17:00:00+00:00                     1   
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                     1   
        $10+ billion (USD) 2024-06-12 05:00:00+00:00                     0   
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00                     1   

  value_salary_est  std_salary_est_perhour  grade_company_revanue       city  \
           70000.0               37.878788                      0  Riverside   
              42.5               42.500000                      0     Newton   
              67.5               67.500000                      0     Remote   
               NaN                     NaN                      9  Cupertino   
              65.0               65.000000                      0  Baltimore   

 country  
      CA  
      MS  
     NaN  
      CA  
      MD

In [265]:
data1['location'].value_counts()['Remote'] #cek beraoa jumlah remote

234

In [266]:
data1['location'].isna().sum() #cek apa ada location yg null

0

In [267]:
data1['country'].isna().sum() #cek apa ada country yg null

335

dari hasil di atas dapat diketahui bahwa jumlah data country yg null > jumlah data remote. Dan data location tidak ada yg null. Artinya, selain remote, ada data location yg hanya berupa city, tidak ada keterangan country nya. Sehingga saya biarkan saja nilai null di country tsb.

##7.Company size dibuat label

In [268]:
data1['company_size'].unique()

array(['501 to 1000 Employees', nan, '51 to 200 Employees',
       '10000+ Employees', '1 to 50 Employees', '1001 to 5000 Employees',
       '5001 to 10000 Employees', '201 to 500 Employees', 'Unknown'],
      dtype=object)

In [269]:
#labeling company_size menjadi level yang dapat diurutkan/dibuat grafik
data1.loc[data1['company_size'] == '1 to 50 Employees', 'level_company_size'] = 1
data1.loc[data1['company_size'] == '51 to 200 Employees', 'level_company_size'] = 2
data1.loc[data1['company_size'] == '201 to 500 Employees', 'level_company_size'] = 3
data1.loc[data1['company_size'] == '501 to 1000 Employees', 'level_company_size'] = 4
data1.loc[data1['company_size'] == '1001 to 5000 Employees', 'level_company_size'] = 5
data1.loc[data1['company_size'] == '5001 to 10000 Employees', 'level_company_size'] = 6
data1.loc[data1['company_size'] == '10000+ Employees', 'level_company_size'] = 7
data1.loc[data1['company_size'] == 'Unknown', 'level_company_size'] = 0
data1.loc[data1['company_size'].isna(), 'level_company_size'] = 0




In [270]:
#ubah level_company_size ke datatype integer
data1['level_company_size'] = data1['level_company_size'].astype('int32')
data1.info()

<class 'pandas.core.frame.DataFrame'>
Index: 1551 entries,  to 
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype              
---  ------                  --------------  -----              
 0   company                 1551 non-null   object             
 1   company_rating          1358 non-null   float64            
 2   location                1551 non-null   object             
 3   job_title               1551 non-null   object             
 4   job_description         1551 non-null   object             
 5   salary_estimate         1274 non-null   object             
 6   company_size            1439 non-null   object             
 7   company_type            1439 non-null   object             
 8   company_sector          1257 non-null   object             
 9   company_industry        1257 non-null   object             
 10  company_founded         1128 non-null   float64            
 11  company_revenue         1439 non-null   object       

In [271]:
data1.head()

company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  company_founded  \
  Information Technology Support Services              NaN   
                                      NaN              NaN   
                Biotech & Pharmaceuticals              NaN   
            Computer Hardware Development           1976.0   
                                      NaN              NaN   

           company_revenue                     dates faktorkali_salary_est  \
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                  1848   
                       NaN 2024-06-11 17:00:00+00:00                     1   
  Unknown / Non-Applicable 2024-06-12 10:00:00+00:00                     1   
        $10+ billion (USD) 2024-06-12 05:00:00+00:00                     0   
  Unknown / Non-Applicable 2024-06-12 04:00:00+00:00                     1   

  value_salary_est  std_salary_est_perhour  grade_company_revanue       city  \
           70000.0               37.878788                      0  Riverside   
              42.5               42.500000                      0     Newton   
              67.5               67.500000                      0     Remote   
               NaN                     NaN                      9  Cupertino   
              65.0               65.000000                      0  Baltimore   

 country  level_company_size  
      CA                   4  
      MS                   0  
     NaN                   2  
      CA                   7  
      MD                   1

##8.Cek apakah Company Type bisa dibikin private n public, kata company dihilangkan

In [272]:
data1['company_type'].unique() #cek dulu unique datanya

array(['Company - Private', nan, 'Company - Public', 'Hospital',
       'Contract', 'Subsidiary or Business Segment',
       'Nonprofit Organization', 'Private Practice / Firm', 'Government',
       'Self-employed', 'College / University', 'Unknown',
       'School / School District'], dtype=object)

karena datanya beragam, maka tidak jadi dibuat boolean antara private dan public. data dibiarkan seperti ini saja.

##9.Melakukan labelling pada company founded

In [273]:
data1['company_founded'].unique()

array([  nan, 1976., 2000., 1998., 1994., 2018., 2019., 1972., 2010.,
       1982., 2003., 2013., 1964., 1969., 2009., 2015., 1984., 1986.,
       1844., 2012., 1999., 1946., 2007., 2020., 2011., 1970., 2016.,
       2017., 1938., 2006., 2001., 2014., 1965., 2022., 1784., 1836.,
       1879., 1926., 1967., 1983., 1849., 1943., 1858., 1916., 1799.,
       1940., 1971., 1980., 1953., 1876., 1992., 1991., 2008., 1851.,
       1928., 1960., 1975., 1973., 1911., 1947., 1892., 1830., 1792.,
       1915., 1995., 1850., 1963., 1959., 1902., 1903., 1918., 1900.,
       1885., 1978., 1899., 1853., 2004., 1989., 2002., 1864., 1955.,
       1898., 1997., 2005., 1968., 1993., 1996., 1901., 2021., 1962.,
       1917., 1942., 1872., 1690., 1877., 1934., 1930., 1921., 1985.,
       1894., 1990., 1875., 1860., 1966., 1936., 1987., 1636., 1910.,
       1852., 1961., 1937., 1939., 1977., 1925., 1841., 1979., 1933.,
       1988., 1948., 1897., 1929., 1810., 1935., 1914., 1913., 1957.,
       1888., 1923.,

In [274]:
#labeling company_founded menjadi bbrp kelompok
data1.loc[data1['company_founded'] < 1900 , 'level_company_founded'] = 'very old'
data1.loc[(data1['company_founded'] >= 1900) & (data1['company_founded'] <= 1999) , 'level_company_founded'] = 'old'
data1.loc[(data1['company_founded'] >= 2000) & (data1['company_founded'] <= 2019) , 'level_company_founded'] = 'middle'
data1.loc[data1['company_founded'] >= 2020 , 'level_company_founded'] = 'new'
data1.loc[data1['company_founded'].isna(), 'level_company_founded'] = 'unknown'
data1.head()



company  company_rating       location  \
               PCS Global Tech\n4.7             4.7  Riverside, CA   
         Futuretech Consultants LLC             NaN     Newton, MS   
                   Clairvoyant\n4.4             4.4         Remote   
                         Apple\n4.2             4.2  Cupertino, CA   
  Skytech Consultancy Services\n5.0             5.0  Baltimore, MD   

                      job_title  \
  Data Engineer | PAID BOOTCAMP   
        Snowflake Data Engineer   
            Data Engineer (MDM)   
                  Data Engineer   
                  Data Engineer   

                                    job_description    salary_estimate  \
  Responsibilities\n· Analyze and organize raw d...  $70000 /yr (est.)   
  My name is Dileep and I am a recruiter at Futu...  $42.50 /hr (est.)   
  Required Skills:\nMust have 5-8+ Years of expe...  $67.50 /hr (est.)   
  Summary\nPosted: Dec 22, 2021\nWeekly Hours: 4...                NaN   
  Description of Work:\nTechnical experience in ...  $65.00 /hr (est.)   

           company_size       company_type                  company_sector  \
  501 to 1000 Employees  Company - Private          Information Technology   
                    NaN                NaN                             NaN   
    51 to 200 Employees  Company - Private  Pharmaceutical & Biotechnology   
       10000+ Employees   Company - Public          Information Technology   
      1 to 50 Employees   Company - Public                             NaN   

                         company_industry  ...           company_revenue  \
  Information Technology Support Services  ...  Unknown / Non-Applicable   
                                      NaN  ...                       NaN   
                Biotech & Pharmaceuticals  ...  Unknown / Non-Applicable   
            Computer Hardware Development  ...        $10+ billion (USD)   
                                      NaN  ...  Unknown / Non-Applicable   

                     dates faktorkali_salary_est value_salary_est  \
 2024-06-12 10:00:00+00:00                  1848          70000.0   
 2024-06-11 17:00:00+00:00                     1             42.5   
 2024-06-12 10:00:00+00:00                     1             67.5   
 2024-06-12 05:00:00+00:00                     0              NaN   
 2024-06-12 04:00:00+00:00                     1             65.0   

  std_salary_est_perhour  grade_company_revanue       city country  \
               37.878788                      0  Riverside      CA   
               42.500000                      0     Newton      MS   
               67.500000                      0     Remote     NaN   
                     NaN                      9  Cupertino      CA   
               65.000000                      0  Baltimore      MD   

 level_company_size  level_company_founded  
                  4                unknown  
                  0                unknown  
                  2                unknown  
                  7                    old  
                  1                unknown  

[5 rows x 21 columns]

#C.Data Demography

##1.Data Statistik berdasarkan describe

In [275]:
data1.describe()

company_rating  company_founded  value_salary_est  \
count     1358.000000      1128.000000       1274.000000   
mean         3.925626      1975.929078      98180.690424   
std          0.535826        51.530079      50629.066214   
min          1.000000      1636.000000         22.500000   
25%          3.600000      1964.750000      84366.000000   
50%          3.900000      1997.000000     102802.000000   
75%          4.200000      2009.000000     125000.000000   
max          5.000000      2022.000000     341000.000000   

       std_salary_est_perhour  grade_company_revanue  level_company_size  
count             1274.000000            1551.000000         1551.000000  
mean                62.644608               3.542876            3.767247  
std                 18.193426               3.610817            2.587436  
min                 22.500000               0.000000            0.000000  
25%                 50.320617               0.000000            1.000000  
50%                 60.000000               3.000000            4.000000  
75%                 71.062229               7.000000            7.000000  
max                233.766234               9.000000            7.000000

1. Rata2 estimasi gaji yang disediakan perusahaan adalah sebesar 62.65 USD per jam
2. Rata2 rating perusahaan adalah 3,93
3. Rata2 grade company revenue adalah 3,5. berdasarkan labeling berarti berada di kisaran 5 sampai 100 juta USD.
4. Rata2 level company size adalah 3,77. Berdasarkan label, berada di kisaran 200 s.d. 1000 karyawannya.


##2.Komposisi proportion level_company_size

In [276]:
#Komposisi Prosentase Level_Company_Size
data1['level_company_size'].value_counts(normalize=True).sort_values(ascending=False)

level_company_size
7    0.279175
2    0.142489
0    0.132173
5    0.130883
1    0.128304
3    0.088975
4    0.054803
6    0.043198
Name: proportion, dtype: float64

Dari data di atas, dapat kita ketahui, perusahaan yg paling banyak melakukan reqruitment adalah perusahaan dg level karyawan tertinggi yaitu level 7 , sebesar 27,9% dari total lowongan (sesuai label, level 7 berarti jumlah karyawan di atas 10 ribu).

##3.Komposisi proportion level_company_founded

In [277]:
# Komposisi berdasarkan company founded
data1['level_company_founded'].value_counts(normalize=True).sort_values(ascending=False)

level_company_founded
old         0.335912
middle      0.315925
unknown     0.272727
very old    0.064475
new         0.010961
Name: proportion, dtype: float64

Dari data di atas dapat diketahui bahwa, job vacancy paling banyak ditawarkan oleh perusahaan OLD (berdasarkan label, didirikan tahun 1900-1999), sebanyak 33,59%. Tidak jauh beda, urutan kedua sebanyak 31,6%  adalah perusahaan MIDDLE (usia menengah, berdasar label didirikan tahun 2000-2019).

##4.Komposisi proportion berdasarkan Tahun

In [278]:
# Komposisi Tanggal reqruitment berdasarkan tahun
data1['dates'].dt.year.value_counts(normalize=True).sort_values(ascending=False)

dates
2024    1.0
Name: proportion, dtype: float64

dari data di atas diketahui bahwa 100 persen reqruitment adalah di tahun 2024.

##5.Komposisi proportion berdasarkan bulan reqruitment

In [279]:
#8. Komposisi tanggal reqruitment berdasarkan bulan
data1['dates'].dt.month.value_counts(normalize=True).sort_values(ascending=False)

dates
6    0.990974
7    0.009026
Name: proportion, dtype: float64

berdasarkan data di atas, 99% reqruitment dilakukan di bulan Juni 2024.

##6.Komposisi proportion berdasrkan country

In [280]:
#Komposisi berdasarkan country
data1['country'].value_counts(normalize=True).sort_values(ascending=False)

country
 CA        0.137336
 TX        0.130757
 VA        0.063322
 NY        0.046875
 GA        0.044408
 MA        0.043586
 IL        0.042763
 NJ        0.041118
 WA        0.036184
 DC        0.035362
 NC        0.034539
 OH        0.032895
 FL        0.032895
 MD        0.028783
 PA        0.026316
 CO        0.021382
 MN        0.019737
 AZ        0.018914
 OR        0.018092
 UT        0.017270
 MO        0.016447
 MI        0.014803
 CT        0.013158
 WI        0.011513
 TN        0.009046
 DE        0.008224
 NV        0.006579
 IA        0.005757
 IN        0.005757
 OK        0.004934
 SC        0.004934
 MS        0.003289
 KY        0.003289
 NE        0.003289
 KS        0.002467
 AR        0.002467
 Bergen    0.001645
 ME        0.001645
 AL        0.001645
 WV        0.001645
 LA        0.001645
 ID        0.000822
 HI        0.000822
 NH        0.000822
 PR        0.000822
Name: proportion, dtype: float64

Berdasarkan data di atas, reqruitment terbanyak dilakukan perusahaan di Negara Bagian CA dan TX dg masing-masing sebesar 13,7% dan 13% dari total reqruitment.